src/ZkFold/Symbolic/Cardano/Types.hs

```diff
@@ -1,13 +1,101 @@
-module ZkFold.Symbolic.Cardano.Types (
-        module ZkFold.Symbolic.Cardano.Types.Address,
-        module ZkFold.Symbolic.Cardano.Types.Script,
-        module ZkFold.Symbolic.Cardano.Types.Tx,
-        module ZkFold.Symbolic.Cardano.Types.TxOut,
-        module ZkFold.Symbolic.Cardano.Types.Value
-    ) where
-
-import           ZkFold.Symbolic.Cardano.Types.Address
-import           ZkFold.Symbolic.Cardano.Types.Script
-import           ZkFold.Symbolic.Cardano.Types.Tx
-import           ZkFold.Symbolic.Cardano.Types.TxOut
-import           ZkFold.Symbolic.Cardano.Types.Value
\ No newline at end of file
+{-# LANGUAGE UndecidableInstances #-}
+module ZkFold.Symbolic.Cardano.Types where
+
+import           Prelude                        hiding ((*), (+), length, splitAt)
+
+import           ZkFold.Symbolic.Data.UInt
+import           ZkFold.Symbolic.Data.UTCTime
+import           ZkFold.Symbolic.Data.ByteString
+import           ZkFold.Base.Algebra.Basic.Number
+import           ZkFold.Base.Data.Vector
+import           ZkFold.Symbolic.Compiler
+
+newtype Transaction inputs rinputs outputs tokens datum a = Transaction
+    ( Vector rinputs (Input datum a)
+    , (Vector inputs (Input datum a)
+    , (Vector outputs (Output tokens a)
+    , (UTCTime a, UTCTime a)
+    ))) deriving Eq
+
+deriving instance
+    ( KnownNat inputs
+    , KnownNat rinputs
+    , KnownNat outputs
+    , KnownNat tokens
+    , Arithmetizable i a
+    , Arithmetizable i (datum a)
+    , Arithmetizable i (UInt 11 a)
+    , Arithmetizable i (UInt 32 a)
+    , Arithmetizable i (UInt 64 a)
+    , Arithmetizable i (ByteString 3 a)
+    , Arithmetizable i (ByteString 8 a)
+    , Arithmetizable i (ByteString 15 a)
+    , Arithmetizable i (ByteString 28 a)
+    ) => Arithmetizable i (Transaction inputs rinputs outputs tokens datum a)
+
+txInputs :: Transaction inputs rinputs outputs tokens datum a -> Vector inputs (Input datum a)
+txInputs (Transaction (_, (is, _))) = is
+
+txOutputs :: Transaction inputs rinputs outputs tokens datum a -> Vector outputs (Output tokens a)
+txOutputs (Transaction (_, (_, (os, _)))) = os
+
+newtype TxId a = TxId a
+    deriving (Eq, Arithmetizable i)
+
+newtype Value n a = Value (Vector n (ByteString 8 a, (ByteString 15 a, UInt 64 a)))
```

Comment:
8 and 15 are the number of bits in `CurrencySymbol` and `TokenName`. respectively. It should be 224 bit for the `CurrencySymbol`, and either 256 or 512 bit for the `TokenName` (please check).

Reply:
@vlasin `TokenName` should be no longer than 32 bytes so it's `256`.